### PR TITLE
Remove result column

### PR DESCRIPTION
**Pull request by BobanL**
# BACKEND PULL REQUEST

## Related Issue

- #3757 
- #4306

## Changes Proposed

- Removes all references to the result (testResult) column and replaces the usage everywhere with Results (set of result)
- Require `Set<Result>` when constructing the Test Event since it is an immutable object 

## Additional Information

- Lazy loading the results may be a better option than eager loading them as they both really experience the same n+1 problem. 
  - `FetchType.LAZY` will generate n queries to get `results` when requested like `testEvent.getResults()`
  - `FetchType.EAGER` will generate n queries  to immediately fetch `results` (but actually does 2n queries
- It is possible to decrease the impact of n queries by setting the batch size by using [`@BatchSize`](https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#fetching-batch) 
- [`Hibernate.initialize`](https://docs.jboss.org/hibernate/orm/4.3/manual/en-US/html_single/#performance-fetching-initialization) is used for forcing data to be fetched before closing the session. This is just a more explanatory way of loading data rather than doing something like `testEvent.getResults` to force the data to be loaded.
  - This generates n queries
  - Our tests run outside of transactions/session so we need to set `hibernate.enable_lazy_load_no_trans` to true in order to fetch results. 
- To avoid n+1 `JOIN FETCH` will need to be applied on every `@Query` where you need the results eagerly. This will only generate one query. ([Example](https://github.com/CDCgov/prime-simplereport/pull/4259/commits/d9b628caa19174630ea6b3de9e5d38e5f2334f4d#diff-3d344713b6d4149b2cf232fdbefc1e60e5d58f868fa04fa4726cf310389e4d12R17-R20))
- You may be able to set [`@Fetch(FetchMode.JOIN)`](https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#fetching-fetchmode-join) on the results field to do a similar action as above. This will eager load the data and should only issue one query, but when I tested this it was issuing n queries in our case.

## Testing

- Smoke test parts of the app that uses test results:
  - test queue
  - results tab
    - correct a result
    - view details
    - print modal
    - download results 

## Checklist for Primary Reviewer

- [ ] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke tested
- [ ] Any content updates (user-facing error messages, etc) have been approved by content team
- [ ] Any changes that might generate questions in the support inbox have been flagged to the support team
- [ ] GraphQL schema changes are backward compatible with older version of the front-end
- [ ] Changes comply with the SimpleReport Style Guide
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [ ] Any dependencies introduced have been vetted and discussed
- [ ] Any changes to the startup configuration have been documented in the README

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java [hunk 1]
```diff
@@ -46,7 +45,10 @@ public class TestEvent extends BaseTestInfo {
   private TestOrder order;
 
   @JsonIgnore
-  @OneToMany(mappedBy = "testEvent", cascade = CascadeType.MERGE, fetch = FetchType.LAZY)
+  @OneToMany(
+      mappedBy = "testEvent",
+      cascade = {CascadeType.MERGE, CascadeType.PERSIST},
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java:50]: ooo! so we needed both here and I didn't realize you could do this! TIL

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java [hunk 2]
```diff
@@ -82,11 +89,15 @@ public TestEvent(TestOrder order, Boolean hasPriorTests) {
       // this can happen during unit tests, but never in prod.
       log.error("Order {} missing PatientAnswers", order.getInternalId());
     }
+    this.results = new HashSet<>();
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java:92]: to confirm... this was to prevent the scenarios where we were initializing `TestEvent`s that had `null` `results`?

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java [hunk 3]
```diff
@@ -94,15 +94,9 @@ public void setDateTestedBackdate(Date date) {
   // - this method is temporary
   // Eventually, this method will be deprecated in favor of getResultSet() and getResultForDisease()
   public TestResult getTestResult() {
-    Hibernate.initialize(this.results);
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java:97]: we are removing this -- is it because as we were seeing this wasn't doing anything to load the `results` on the `TestOrder`?

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/BaseTestInfo.java [hunk 4]
```diff
@@ -39,11 +38,6 @@ public abstract class BaseTestInfo extends AuditedEntity implements Organization
   @JoinColumn(name = "specimen_type_id")
   private SpecimenType specimenType;
 
-  @Column(nullable = true)
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/BaseTestInfo.java:42]: I remember discussing this at one point. How come we didn't add a `results` column here since this is a parent class for `TestEvent` and `TestOrder`? 🤔 Curious for my own Java leaning purposes 😅

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/TestOrderRepository.java [hunk 5]
```diff
@@ -16,14 +16,15 @@
...
       "from #{#entityName} q "
+          + "left join fetch q.results "
           + "where q.organization.isDeleted = false "
           + "and q.patient.isDeleted = false ";
   String BASE_ORG_QUERY = BASE_QUERY + " and q.organization = :org ";
   String FACILITY_QUERY = BASE_ORG_QUERY + " and q.facility = :facility ";
   String IS_PENDING = " and q.orderStatus = 'PENDING' ";
   String IS_COMPLETED = " and q.orderStatus = 'COMPLETED' ";
   String ORDER_CREATION_ORDER = " order by q.createdAt ";
-  String RESULT_RECENT_ORDER = " order by updatedAt desc ";
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/repository/TestOrderRepository.java:26]: haha so was this working before?

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java [hunk 6]
```diff
@@ -156,20 +172,10 @@ public DeviceSpecimenType getDeviceSpecimenType() {
...
-      if (resultObject.isPresent()) {
-        return resultObject.get().getTestResult();
-      }
-    }
-    return super.getResult();
-  }
-
-  @Override
-  public TestResult getResult() {
-    return getTestResult();
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java:173]: I'm wondering if we should update the name of this method to be `getCovidTestResult` so we can be more clear about what result we are getting back?

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java (hunk 3 above)
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java:97]: So after looking into this more this morning, there are two approaches to using `Hibernate.initialize` or Eager/Join Fetch. `Hibernate.initialize` will issue an additional query to fetch the data for results while Eager/Join Fetch will fetch that data in just one query. 

I don't think that result set is a very heavy object so an eager load shouldn't impact performance and we already seem to use result in a lot of places we get TestEvent or TestOrder so we are already fetching that data with a second query.

File: backend/src/main/resources/graphql/main.graphqls [hunk 7]
```diff
@@ -237,7 +237,6 @@ type TestOrder {
   symptomOnset: LocalDate
   deviceType: DeviceType
   deviceSpecimenType: DeviceSpecimenType
-  result: String
```
Comment (emyl3) [inline backend/src/main/resources/graphql/main.graphqls:240]: learned from @emmastephenson that we can use  `@deprecated(reason: "use results as source of truth")` or something like that here instead of removing them. I can go ahead and clean this up!

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/TestOrderRepository.java (hunk 5 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/repository/TestOrderRepository.java:26]: Got it! Thank you!

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java (hunk 6 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java:173]: Thank you so much for doing this -- it is so much easier follow!! 😸

File: backend/src/main/resources/graphql/main.graphqls (hunk 7 above)
Comment (emyl3) [reply backend/src/main/resources/graphql/main.graphqls:240]: Done here [65ba33](https://github.com/CDCgov/prime-simplereport/pull/4259/commits/65ba3365f17de52783ccf85b5ddf99b94131f329)

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/BaseTestInfo.java [hunk 8]
```diff
@@ -39,11 +38,6 @@ public abstract class BaseTestInfo extends AuditedEntity implements Organization
...
-  @Type(type = "pg_enum")
-  @Enumerated(EnumType.STRING)
-  private TestResult result;
```
Comment (BobanL) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/BaseTestInfo.java:45]: Should we be deleting this field or just like deprecating it?

Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/BaseTestInfo.java:45]: I think just deprecating it - the `result` column is still on the schema definition, and I'm not sure how Java/liquibase will response if the schema definition doesn't exactly match the Spring models.

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java [hunk 9]
```diff
@@ -94,21 +93,9 @@ public void setDateTestedBackdate(Date date) {
...
-      Optional<Result> resultObject = this.results.stream().max(resultDateComparator);
-      if (resultObject.isPresent()) {
-        return resultObject.get().getTestResult();
-      }
-    }
-    return super.getResult();
-  }
-
-  @JsonIgnore
-  public Set<Result> getResultSet() {
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java:109]: (non-blocking/question for my own knowledge) When would we use this method as opposed to `getResults()`?

**Review by emyl3 — APPROVED** [review]
This looks good to me! I smoke tested the app locally and all test result related actions worked for me as intended. Thank you for helping me with this ticket 😭 🙌 🙏 😹

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java (hunk 9 above)
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java:109]: This method is only used to resolve the field [`TestOrder.TestResults`](https://github.com/CDCgov/prime-simplereport/blob/main/backend/src/main/java/gov/cdc/usds/simplereport/api/queue/ApiTestOrderDataResolver.java#L103-L107). I just put an `@Deprecated` on this method so no one uses this by accident in the future

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/BaseTestInfo.java (hunk 8 above)
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/BaseTestInfo.java]: Ok yeah, I didn't know if it would screw something up! Just added it back and slapped a `@Deprecated` on it.

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/Result.java [hunk 10]
```diff
@@ -58,8 +58,7 @@ public Result(TestOrder testOrder, SupportedDisease disease, TestResult testResu
   }
 
   /* Copy constructor, used for corrections and removals */
-  public Result(Result originalResult, TestEvent testEvent) {
-    this.testEvent = testEvent;
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/Result.java:62]: Why are we removing this testEvent, it seems unrelated to the other `result` changes?

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java [hunk 11]
```diff
@@ -154,22 +168,12 @@ public DeviceSpecimenType getDeviceSpecimenType() {
 
   // This method is temporary and eventually, this method will be deprecated in favor of
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java:169]: This comment about the temporary method can probably be removed!

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java [hunk 12]
```diff
@@ -154,22 +168,12 @@ public DeviceSpecimenType getDeviceSpecimenType() {
...
-  }
-
-  @Override
-  public TestResult getResult() {
-    return getTestResult();
+    Optional<Result> resultObject =
+        this.results.stream()
+            .filter(result -> COVID_LOINC.equals(result.getDisease().getLoinc()))
+            .findFirst();
+    return resultObject.map(Result::getTestResult).orElse(null);
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestEvent.java:177]: Hmmm, I don't love the possibility of returning `null` on this method...I think we could either return `Optional<TestResult>`, or just throw an exception if someone tries to call this method on a TestEvent that doesn't have a covid result. But nulls tend to bubble up in sneaky ways.

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java (hunk 9 above)
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java:109]: Given that the `results` field is also deprecated in the `TestOrder` on `main.graphqls`, I think it should be safe to remove the reference from `ApiTestOrder` as well. Then we can remove this `getTestResult` method entirely, instead of just deprecating it. (I just tested locally, and graphQL doesn't blow up when you remove the deprecated `result` out of `ApiTestOrder` - I think at that point, it's the responsibility of the client not to call it). 
I'd personally feel a bit better about removing this method entirely, because the possibility of returning null makes me nervous 😅

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java [hunk 13]
```diff
@@ -416,32 +421,32 @@ public AddTestResultResponse addMultiplexResult(
...
-        });
+    return newResults.stream()
+        .map(
+            newResult -> {
+              Optional<Result> pendingResult =
+                  _resultRepo.getPendingResult(
+                      order, _diseaseService.getDiseaseByName(newResult.getDiseaseName()));
+              if (pendingResult.isPresent()) {
+                pendingResult.get().setResult(newResult.getTestResult());
+                order.addResult(pendingResult.get());
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:435]: Does this just get us around the initialization issue with `order.getResults()`? `_resultRepo.save(pendingResult)` should be enough to persist the result in the database with a link to the testOrder, so I'm curious what this extra call does.

**Review by emmastephenson — COMMENTED** [review]
Overall this is looking solid, thanks Boban & Elisa! 

I have a couple questions about the model updates, if it's easier to hop on a call to talk through those just let me know.

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java (hunk 9 above)
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java:109]: Removed it in this commit ebc34c9859a314bc210370ca938d55b30d474451

**Review by georgehager — COMMENTED** [review]
Overall looks great, since I am not as familiar with this code I don't have a lot of comments to add.

I've pulled the branch locally and have gone through smoke testing, all the flows look good except I'm seeing an issue with corrected results for multiplex. It appears that when trying to correct a multiplex result, it errors out because it's still looking for `result`. Here are the screenshots:
Screen Shot 2022-09-26 at 10 56 15 AM

Screen Shot 2022-09-26 at 10 55 41 AM

Screen Shot 2022-09-26 at 10 54 01 AM

Test correction works fine on covid-only results.

I've got the feature flag enabled for multiplex locally, so I don't think that's the issue, but I'll try to double check that it's not caused by my local configuration.

Another thing, I'm seeing the same issue that @BobanL and @emyl3 flagged the other day with results initially not displaying correctly:

Screen Shot 2022-09-26 at 10 48 36 AM

Once I added another result with multiplex it displayed fine:

Screen Shot 2022-09-26 at 10 51 59 AM

I'm guessing this will be fine in environments, but haven't verified.

**Comment by BobanL** [comment]
@georgehager I think most of those issues were caused because I did not merge main into this branch. I did that and redeployed to dev for you to test again!

**Review by emmastephenson — APPROVED** [review]
LGTM, thanks for digging into the JPA oddities on this one Boban! Tested in `dev` and everything looks great.

**Review by georgehager — APPROVED** [review]
